### PR TITLE
Dataproc metastore - add missing args

### PR DESCRIPTION
**Pull request by DrFaust92**
Closes https://github.com/hashicorp/terraform-provider-google/issues/10568

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
metastore: add `endpoint_protocol`,  `database_type`, `release_channel `, `network_config` arguments and `uid` attribute to `dataproc_metastore_service`
```

### CODE REVIEW

**Comment by Noremac201** [comment]
Thanks for doing this I was actually in the middle of adding PSC support, but I'll let you take over.

A minor nit is that some of these are preview only like `networkConfig`, can you add `min_version: beta` to that?

File: mmv1/products/metastore/api.yaml [hunk 1]
```diff
@@ -188,3 +189,55 @@ objects:
...
+        values:
+          - :CANARY
+          - :STABLE
+      - !ruby/object:Api::Type::String
+        name: 'uid'
+        output: true
+        description: |
+          The globally unique resource identifier of the metastore service.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'networkConfig'
```
Comment (Noremac201) [inline mmv1/products/metastore/api.yaml:221]: Can you add `min_version: beta` here?

**Review by Noremac201 — CHANGES_REQUESTED** [review]

File: mmv1/products/metastore/api.yaml [hunk 2]
```diff
@@ -188,3 +189,55 @@ objects:
...
+            name: 'consumers'
+            description: |
+              The consumer-side network configuration for the Dataproc Metastore instance.
+            properties:
+              - !ruby/object:Api::Type::String
+                name: 'endpointUri'
+                output: true
+                description: |
+                  The URI of the endpoint used to access the metastore service.
+              - !ruby/object:Api::Type::String
```
[the remark this answers is not in the document]
Comment (nitikhadapkar) [reply mmv1/products/metastore/api.yaml:235]: Network field is internal, theres no support for it and it should not be exposed.

File: mmv1/products/metastore/api.yaml (hunk 1 above)
Comment (DrFaust92) [reply mmv1/products/metastore/api.yaml:221]: Done, thanks!

**Comment by slevenick** [comment]
Is this ready for review? It's marked draft still

**Comment by DrFaust92** [comment]
Basic test passes but I haven’t fully tested all new attributes yet. If
there are any comments meanwhile that’s great as well 😊

**Review by slevenick — CHANGES_REQUESTED** [review]
The changes look reasonable, but can you add a test that uses these settings? Especially the `consumers` block, but ideally all the new fields

**Comment by slevenick** [comment]
Looks like the Dataproc Metastore test fails with:
`    provider_test.go:309: Step 1/2 error: Check failed: Check 1/1 error: google_dataproc_metastore_service.my_metastore: Attribute 'network_config.0.consumers.0.endpoint_uri' expected to be set`

Are you able to run this test successfully locally?

**Comment by DrFaust92** [comment]
[slevenick](https://github.com/slevenick) actually having trouble with this locally i gave it a shot here to check if its an issue on my env.

**Comment by DrFaust92** [comment]
Interesting error on server side when i run locally the metastore tries to create, times out and deletes itself.

when looking at GCP logs explorer i find this:

```
peration failed with error: 
generic::invalid_argument: terraform apply failed, error: exit status 1, stderr: 
2022/04/15 14:54:22 [DEBUG] Using modified User-Agent: Terraform/0.12.31 Cloud SSA

Error: Error waiting to create Connector: Error waiting for Creating Connector: Error code 13, message: An internal error occurred: VPC Access connector failed to get healthy. Please check GCE quotas, logs and org policies and recreate.

  on main.tf line 68, in resource "google_vpc_access_connector" "vpc_connector":
  68: resource "google_vpc_access_connector" "vpc_connector" {



, stdout: 
```

**Comment by slevenick** [comment]
Hmmm now it's getting the internal error 13. Not sure why that is happening, maybe a transient error in the service?

**Comment by DrFaust92** [comment]
Ive opened an internal ticket to investigate this as either there is so pre-req missing or https://cloud.google.com/dataproc-metastore/docs/psc-ilb#gcloud is wrong.

**Comment by Noremac201** [comment]
I looked into it, seems to be a transient issue on our side, a retry should work

**Comment by slevenick** [comment]
Yikes, this is worrying!

```
=== CONT  TestAccDataprocMetastoreService_networking
panic: interface conversion: interface {} is []interface {}, not map[string]interface {}

goroutine 585 [running]:
github.com/hashicorp/terraform-provider-google-beta/google-beta.flattenDataprocMetastoreServiceNetworkConfigConsumers(...)
	/go/src/github.com/hashicorp/terraform-provider-google-beta/google-beta/resource_dataproc_metastore_service.go:885
```

File: mmv1/products/metastore/api.yaml [hunk 3]
```diff
@@ -203,3 +204,60 @@ objects:
...
+        output: true
+        description: |
+          The globally unique resource identifier of the metastore service.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'networkConfig'
+        min_version: beta
+        description: |
+          The configuration specifying the network settings for the Dataproc Metastore service.
+        properties:
+          - !ruby/object:Api::Type::NestedObject
```
Comment (Noremac201) [inline mmv1/products/metastore/api.yaml:244]: This is a list of NestedObjects
```
- !ruby/object:Api::Type::Array
  name: 'consumers'
  item_type: !ruby/object:Api::Type::NestedObject
```

Comment (DrFaust92) [reply mmv1/products/metastore/api.yaml]: Thanks, fixed!

**Comment by DrFaust92** [comment]
[slevenick](https://github.com/slevenick) I couldn't test earlier because of the transient service error above. but i can replicate this error. from what i can gather the metastore creates an ip address implicitly (and also removes it that way)

I think we need to add some retry on the subnet. but i wonder if its a bit out of scope for this PR. I can look into that and add some best effort 1/2 min retry.

File: mmv1/third_party/terraform/tests/resource_dataproc_metastore_service_test.go.erb [hunk 4]
```diff
@@ -53,6 +81,39 @@ resource "google_dataproc_metastore_service" "my_metastore" {
...
+	service_id = "%[1]s"
+	location   = "us-central1"
+	tier       = "%[2]s"
+
+	hive_metastore_config {
+		version = "2.3.6"
+	}
+
+	network_config {
+      consumers {
```
Comment (slevenick) [inline mmv1/third_party/terraform/tests/resource_dataproc_metastore_service_test.go.erb:110]: Spacing here needs fixing

**Comment by slevenick** [comment]
> [slevenick](https://github.com/slevenick) I couldn't test earlier because of the transient service error above. but i can replicate this error. from what i can gather the metastore creates an ip address implicitly (and also removes it that way)
> 
> I think we need to add some retry on the subnet deletion. but i wonder if its a bit out of scope for this PR. I can look into that and add some best effort 1/2 min retry.

I don't think the problem here is on the subnetwork side. It seems like metastore creates the ip address but doesn't completely remove it before considering itself deleted, which makes Terraform attempt to delete the subnetwork before the dependency is removed.

I believe this is a metastore bug, as it should fully clean up itself and any subresources before considering itself finished deleting

File: mmv1/third_party/terraform/tests/resource_dataproc_metastore_service_test.go.erb (hunk 4 above)
Comment (DrFaust92) [reply mmv1/third_party/terraform/tests/resource_dataproc_metastore_service_test.go.erb]: done

**Comment by DrFaust92** [comment]
[slevenick](https://github.com/slevenick), any thing i can do from my side?

**Comment by slevenick** [comment]
@Noremac201 do you have any insight on the above issue with cleaning up the metastore IP address?

**Comment by Noremac201** [comment]
As part of deletion, we "fast-delete" the service and there is a grace period where we'd re-use all existing infrastructure, so it _is_ cleaned up eventually (the address) but not in the timespan of this test.

Spinning up a new network, creating a metastore, and then removing all of them is kind of an edge case, I'll bring it to the team to see how we want to tackle this.

However, for now, this is unfortunately expected.

**Comment by DrFaust92** [comment]
Sorry [slevenick](https://github.com/slevenick), was a way for a bit. will try it now

**Comment by DrFaust92** [comment]
@slevenick, so i tired splitting the test config into 2 as suggested. it did not help. ill see if a i can use some default subnet to at least appease tests. Any example of doing this in any other test config i can rely on?

**Comment by DrFaust92** [comment]
@slevenick how would you feel if i add something like this? https://github.com/hashicorp/terraform-provider-aws/blob/6e8fe78d5d52229122afc6748d6ff021c4409ab3/internal/service/secretsmanager/secrets_data_source_test.go#L21-L50

**Comment by DrFaust92** [comment]
[Noremac201](https://github.com/Noremac201) any thoughts on the above?

**Comment by DrFaust92** [comment]
[slevenick](https://github.com/slevenick) / [Noremac201](https://github.com/Noremac201) Ive opted to use a datasource for default network, locally tests pass. i hope not it will pass in CI

**Comment by zli82016** [comment]
Hello, @DrFaust92 , do you have the plan to continue working on this PR? Thanks.

**Comment by DrFaust92** [comment]
These should be supported already

**Comment by zli82016** [comment]
Thanks for your response. I checked the code. Yes, only the field `network_config` is not supported. Is it because the corner case which caused the test failing?

**Pull request closed without merging**